core/shared/src/main/scala/zio/ZIOMetric.scala

```diff
@@ -263,6 +264,25 @@ object ZIOMetric {
     final def increment(implicit trace: ZTraceElement): UIO[Any] =
       counter.increment(1.0)
 
+    /**
+     * Converts this counter metric to one where the tags depend on the measured
+     * effect's result value
+     */
+    def withDynamicTags(f: A => Chunk[MetricLabel]): ZIOMetric[A] =
+      new ZIOMetric[A] {
+        override def apply[R, E, A1 <: A](zio: ZIO[R, E, A1])(implicit trace: ZTraceElement): ZIO[R, E, A1] =
+          self.apply(zio.tap(changeCounter))
+
+        private def changeCounter(value: A)(implicit trace: ZTraceElement): UIO[Unit] =
+          ZIO.succeed {
+            val extraTags = f(value)
+            val allTags   = self.tags ++ extraTags
```

Comment:
it could also replace the tags instead of combining the static ones with the dynamic ones